text/3712-box-yesalias.md

```diff
@@ -0,0 +1,127 @@
+
+Any `unsafe` code that may want to temporarily maintain aliased `Box<T>`s for various reasons (such as low-level copy operations), or may want to use something like `ManuallyDrop<Box<T>>`, is put into an interesting position: While they can use a user-defined smart pointer, this requires both care on the part of the smart pointer implementor, but also affects the ergonomics and expressiveness of that code, as `Box<T>` has many special language features that surface at the syntax level, which cannot be replicated today by a user-defined type.
+
+[^2]: It is possible to write a type that has the same issue in theory, involving a `&UnsafeCell<T>` field and manually holding an a pointer to an allocation. Holding any other type of reference, however, would cause immediate UB if a function parameter with that type was dropped, due to the strong protector, compared to a weak protector used for `Box<T>`, which allows deallocation, and no protector used for `&UnsafeCell<T>`. For the same reasons as `Box<T>` here, I would personally discourage any type from directly holding a reference that is invalidated in its `Drop`, and for the most part, this is protected against by the borrow checker, and requires unsafe lifetime extension to achieve.
+[^3]: Since [rust#122018](https://github.com/rust-lang/rust/pull/122018), this is fixed as only `Box<T>` has the special opsem behaviour described here, which is the second solution.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+[`alloc::boxed::Box<T>`] is defined with a number of magic features that interact directly with the language. To limit the impact, both on users and on the language itself, we remove the language level requirement that `Box` be dereferenceable and unique (like applies to `&mut T`). 
+This does not affect library level requirements of uniqueness however - it may remain undefined behaviour to access a `Box` that is being aliased (particularily where the operations produce multiple aliasing mutable references from different `Box`es) or to use `Box::from_raw` to construct multiple aliasing boxes. 
+
+We also do not remove any other language level magic from `Box`, such as the ability to do both partial and complete moves from a `Box`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+For the remainder of this section, let `WellFormed<T>` designate a type for *exposition purposes only* defined as follows:
+```rust
+#[repr(transparent)]
+struct WellFormed<T: ?Sized>(core::ptr::NonNull<T>);
+```
```

Comment:
Are there any differences semantically between what is proposed for `Box<_>` in this RFC and what would be true of `MaybeDangling<Box<_>>` today?

That is, since we already accepted [RFC 3336](https://github.com/rust-lang/rfcs/pull/3336), if there is no daylight between these, then it should say that normatively, and it perhaps could even lean into that for defining the semantics.

cc @RalfJung

Reply:
We haven't spelled out whether `MaybeDangling<Box<_>>` would allow pointers too close to the end of the address space, but it seems reasonable to say "no" to that. In that case the answer to your question is yes, this RFC proposes to weaken `Box` so that its validity requirements are equivalent to `MaybeDangling<Box<_>>`.

Reply:
> Are there any differences semantically between what is proposed for `Box<_>` in this RFC and what would be true of `MaybeDangling<Box<_>>` today?

Restating Ralf's answer to this, `MaybeDangling` *only* removes aliasing invariants, but preserves the normal validity invariants of the type. This RFC proposes to remove the aliasing invariants from `Box<_>` as a whole, so it would naturally leave it in an identical state to `MaybeDangling<Box<_>>`. 

However, this also spells out that validity invariant in full, which we cannot rely on existing types yet to do.